### DISCUSSION TITLE
Transparency issue when rotating .PNGs

### DISCUSSION THREAD

**Discussion by EvilSupahFly** [ImageMagick/ImageMagick#6765] [Help] [answered]
### ImageMagick version

6.9.11-60 Q16 x86_64 2021-01-25

### Operating system, version and so on

Linux Mint 21.2 Cinnamon, Kernel 6.4.0-060400-generic x64, Intel© Core™ i7-4770S CPU x 4 @ 3.10GHz

### Description

Yes, I know it's not the newest and greatest version - it's just the one that comes from my distro's repo.

When I rotate a PNG with existing transparency, 45 degrees for example, the overall image is enlarged and the "corners" are painted in a solid colour (which you can see in the attached screenshot, showing the original, and a version rotated 45 degrees), rather than extending the alpha channel and giving me complete background transparency. I've tried all the "-alpha [xxx]" options but nothing gives me the result I'm looking for.

Sure, I could open the new image, fix the opaque corners, and trim the image back down to the original size, but I wrote a script that rotates the original image in 5 degree increments intending to make a custom boot screen animation for Plymouth - meaning manually fixing seventy two PNG files, which I would rather not have to do.

Is there something I'm misunderstanding, or doing wrong? Or is this just how it is?

![Screenshot from 2023-10-06 18-08-46](https://github.com/ImageMagick/ImageMagick/assets/3280846/9affdfc3-7de1-4af1-934c-6d83d0463cc4)

**Comment by fmw42** [comment]
What was your exact command line?  Does it work in a single command line outside your script? Post the original image by itself.  Are you sure you are using that particular version of Imagemagick?  It might be a bug in that version.  I suggest you upgrade to IM 7 if possible.

**Comment by snibgo** [comment]
If you are using `-rotate`, see the documentation https://imagemagick.org/script/command-line-options.php#rotate

> Empty triangles in the corners, left over from rotating the image, are filled with the `-background` color.

**Comment by EvilSupahFly** [comment]
`convert "logo.png" -rotate 45 "logo_001.png"` is the manual rotation I used for the screenshot. I also cycled through all the "-alpha" options sequentially, but still had the triangles in each corner because of how `-rotate` works, which doesn't seem to be different in any way from the current stable available directly on the webpage, and the slightly older version in my distro's repo.

I tried changing the background colour to the same as the alpha channel in order to achieve the transparency effect I wanted, but that didn't work either as `-background` requires a colour name or hex code, and doesn't accept "transparent" or "alpha":

> https://imagemagick.org/script/command-line-options.php#background:
> The color is specified using the format described under the [-fill](https://imagemagick.org/script/command-line-options.php#fill) option. The default background color (if none is specified or found in the image) is white.
> ```-fill blue```
> ```-fill "#ddddff"```
> ```-fill "rgb(255,255,255)" ```

I assume, then, I have to modify the original file in some way as to make the background colour something not already in use by the original image, which I can pass to `-transparent` instead of trying to use `-alpha`, and since it appears that my original image uses `#000000` for the alpha channel, if I pass that during the rotation process, the visible black areas will be converted to transparent pixels, which will ruin the image. So, my guess is that if I change the alpha channel to `#FF00FF` which I believe I can do in GIMP, and pass `-fill "#FF00FF" -transparent "#FF00FF"` it will make the background transparent, right?

Also assuming here that a little work with `-geometry` should preserve the image size? Although, the perception of the logo shrinking and growing as it spins while the system loads is appealing, so I may not play with `-geometry`.

For the record, this is my original logo file, followed by the rotated copy.

![logo](https://github.com/ImageMagick/ImageMagick/assets/3280846/54e78e99-ac24-4fb3-8775-781605ec962b)
![logo_001](https://github.com/ImageMagick/ImageMagick/assets/3280846/bd98cc11-dbea-4a66-b3d5-bca89bd8025b)

**Comment by EvilSupahFly** [comment]
Well, tweaking the transparency colour didn't have the effect I wanted, and only messed up the picture. Perhaps this was just a bad idea?
![logo_modified_alpha](https://github.com/ImageMagick/ImageMagick/assets/3280846/971aec1c-6081-4a5c-807f-e038b4350c6c)

**Comment by snibgo** [comment, marked as answer]
> -background requires a colour name or hex code, and doesn't accept "transparent" or "alpha"

It does accept `transparent`. Like this:

```
magick "logo.png" -background transparent -rotate 45 +repage "logo_001.png"
```
Or you can use `-background None` which means the same thing.

**Reply by EvilSupahFly** [reply]
This resolved my transparency issue quite nicely.

**Comment by EvilSupahFly** [comment]
> > -background requires a colour name or hex code, and doesn't accept "transparent" or "alpha"
> 
> It does accept `transparent`. Like this:
> 
> ```
> magick "logo.png" -background transparent -rotate 45 +repage "logo_001.png"
> ```
> 
> Or you can use `-background None` which means the same thing.

That was what I was looking for! In my script, I wrote the loop to step by 5 so I could pass the iteration tracker as the rotation argument, since I wanted it to be automatic (knowing there would be 72 "frames" for the animation):
```
for i in {5..360..5}; do
    ((c=c+1))
    if (( c < 10 )); then
        fileOut="$fileBase""00""$c.$fileExt"
    elif (( c > 9 && c < 100 )); then
        fileOut="$fileBase""0""$c.$fileExt"
    else
        fileOut="$fileBase$c.$fileExt"
    fi
    convert "$fileIn"-rotate $i -background transparent +repage "$fileOut"
    sleep 2
done
```
With transparency resolved, I would guess that I have to use `-geometry` to increase the image size by the percentage difference between the original logo and each of the 71 rotated frames, then crop out the extra space on the edges.

In my subsequent testing, I've determined that the percentage of increase isn't consistent with each 5 degree rotational increment, which makes me believe I would have to have a second loop to process each new frame, compare it to the original file, determine the difference  and trim that difference from each of the four edges - preserving the original centre point so the illusion of animation isn't spoiled by wobbling.

I know `-geometry` can crop using coordinates for a specific offset, and will preserve the aspect ratio (1:1 in my case), but each frame will have a different set of offset coordinates, although, if my math is correct, then theoretically, I should only need 9 sets of coordinates, right? Each 90 degree increment after 45 should be the biggest the images get: 45 should be the same as 135, 225, and 315 requiring the most trim while 90, and 270 should be the same size as the original, needing no trim.

Between 5 and 45 degrees, the frame size grows. From 50 to 90, it shrinks by the same percentage sequence it grew by, just in reverse, with the 90 degree version being exactly the same size as the original. From 95 to 135 it grows again, and so on. It's a square so each increase will have a corresponding decrease. The problem is that's a lot of dynamic math for a simple BASH script.

I suppose I could run the first set of 9, and manually determine each percentage and subsequent offset location, then code those locations into the loop in an array, knowing that I have to cycle forward and backward through the array every 9 files, but surely there's a better way?

I thought it would be just a simple thing: write a script to take a file as input, make a loop that steps by +5, and rotate the image that many degrees and save it as a new incrementally named file. ImageMagick can't determine the percentage differences and subsequent offset locations by itself., right? So maybe I should just do this in Python or something instead of a shell script?

**Comment by snibgo** [comment]
I suggest upgrading to IM v7 if you can. The v7 syntax is more powerful and more logical, but is also stricter, executing options in the command line in the same order that you give them. So with v7 your command would be:
```
magick "$fileIn" -background transparent -rotate $i +repage "$fileOut"
```
Note that `-background` needs to be before `-rotate`.

If you need all the images to be the same size, one obvious method is to `-extent` them all to that size. For a square input, the maximum size is at 45 degrees as you say, and the size (new width and height) is `floor(W*sqrt(2)+2)` where W is the input width and height. The `+2` is to allow for antialiasing. You could calculate that once (eg with `bc`) and use that like this:
```
magick "$fileIn" -background transparent -rotate $i +repage -gravity Center -extent ${NewDim}x${NewDim} "$fileOut"
```

**Reply by EvilSupahFly** [reply]
Math was never my strength (and apparently I fail at Googling too lol). IM has literally so many options - I should have upgraded in the first place and taken more time to read the documentation. Lesson learned, right? This is fantastic help, however, so I'm extremely appreciative of your feedback. I'm in the process now (albeit remotely, via my phone) of performing a few software updates and upgrades, and I'm adding IM to the queue, so when I get home from work this evening, I'll try this again.

Thanks a tonne!

Edit: Adding a buffer zone to accommodate AA is a **really nice** touch - I wouldn't have thought of that.

**Reply by EvilSupahFly** [reply]
You sir, have solved literally everything! Here's the original logo, and here's the 45 degree rotation using your help!

![logo](https://github.com/ImageMagick/ImageMagick/assets/3280846/89dcea84-61a6-452a-8f0a-8fe5ab747a00)
![logo_test](https://github.com/ImageMagick/ImageMagick/assets/3280846/17752fad-f1e3-4129-8259-711e3ff19354)

**Comment by snibgo** [comment]
Good stuff.

For complex transformations for animation, "`-distort SRT`" can be very useful. See [Animation with SRT](http://im.snibgo.com/animsrt.htm) . It isn't needed in your simple case.

**Reply by EvilSupahFly** [reply]
Indeed, since Plymouth simply cycles through a series of static image files, but that's good to know should I want to take on something more complex.

**Comment by snibgo** [comment]
We can do the work in a single v7 `magick` command. I break it into lines for legibility.
```
magick \
  toes.png \
  -duplicate 71 \
  -background transparent \
  -set option:NewDim %[fx:hypot(w,h)+2] \
  -rotate %[fx:t*5+5] +repage \
  -gravity Center -extent %[NewDim]x%[NewDim] \
  -write myframes-%03d.png \
  -set dispose Background x.gif  
```
This makes frame files named `myframes-000.png` etc, and a single animated GIF.

It works by making 71 duplicates of the input. It sets `NewDim` to `hypot(w,h)+2`, the diagonal of the input image, plus 2. Then it rotates each of the 72 images by `t*5+5` degrees, where `t` is the index of that image in the list (0, 1, 2, ...). Then it does the extent, then it writes to files named `myframes-000.png`, `myframes-001.png`, etc. Then it also writes an animated GIF.

This will be faster than using a shell loop as we don't need to read the input file 72 times. But it needs more memory.

**Reply by EvilSupahFly** [reply]
Interesting. I see it also increments the angle and creates sequentially numbered files too.

I could drop that into a shell script instead, and pass a target file on the command line, or I could make it a function in `.bashrc` like I've done for several ffmpeg functions I use frequently so I can use it wherever, whenever... I could make the the input file and the rotation passable arguments in both cases too, which increases the flexibility!

This is outstanding! The more I learn about ImageMagick, the more I'm impressed by it!
